src/collection.jl

```diff
@@ -115,30 +194,198 @@ end
-    for row in eachrow(df)
-        @yield FITS(row.path)[row.hdu]
+# generator for CCDData specified by data frame (i.e. path of file, hdu etc.)
+@resumable function ccds(collection)
+    for row in eachrow(collection)
+        @yield CCDData(row.path; hdu = row.hdu)
     end
 end
+
+
+"""
+    ccds(f, collection; save = any(!isnothing, (save_prefix, path, save_suffix)), path = nothing, save_prefix = nothing, save_suffix = nothing, save_delim = "_", ext = r"fits(\\.tar\\.gz)?"i, kwargs...)
+
+Iterates over the `CCDData`s of the collection applying function `f` at each step.
+
+It returns an array of output values of function `f` applied on `CCDData`s.
+In addition to applying function `f`, the outputs can be saved. If `save = true`, it enables programmatical saving of returned value of the function `f` using [`CCDReduction.writefits`](@ref). File is saved at `path` specified by the user.
+Suffix and prefix can be added to filename of newly created files by modifying `save_suffix` and `save_prefix`, `save_delim` is used as delimiter.
+`ext` is the extension of files in collection, by default it is set to `r"fits(\\.tar\\.gz)?"i`.
+
+# Example
+```julia
+collection = fitscollection("~/data/tekdata")
+processed_images = map(ccds(collection)) do img
+    trim(img, (:, 1040:1059))
+end
+```
+The above generates `processed_images` which consists of trimmed versions of images present in `collection`.
+
+For saving the `processed_images` simultaneously with the operations performed
+```julia
+processed_images = map(ccds(collection; save = true, path = "~/data/tekdata", save_prefix = "trimmed")) do img
+    trim(img, (:, 1040:1059))
+end
+```
+The trimmed images are saved as `trimmed_(original_name)` (FITS files) at `path = "~/data/tekdata"` as specified by the user.
+
+Mapping version of `ccds` function is interfaced on iterative version of `images`, any valid parameter can be passed into iterative version as `kwargs`.
+"""
+function ccds(f, collection; save = any(!isnothing, (save_prefix, path, save_suffix)), path = nothing, save_prefix = nothing, save_suffix = nothing, save_delim = "_", ext = r"fits(\.tar\.gz)?"i, kwargs...)
+    image_iterator = ccds(collection; kwargs...)
+    locations = collection.path
+
+    processed_images = map(zip(locations, image_iterator)) do (location, output)
+        processed_image = f(output)
+        if save
+            save_path = generate_filename(location, path, save_prefix, save_suffix, save_delim, ext)
+            writefits(save_path, processed_image)
+        end
+        processed_image
+    end
+
+    return processed_images
+end
+
+
+"""
+    filenames(f, collection; save = any(!isnothing, (save_prefix, path, save_suffix)), path = nothing, save_prefix = nothing, save_suffix = nothing, save_delim = "_", ext = r"fits(\\.tar\\.gz)?"i, kwargs...)
+
+Iterates over the file paths of the collection applying function `f` at each step.
+
+It returns an array of output values of function `f` applied on file paths.
```

Comment:
```suggestion
Loads the data associated with each row of `collection` into an `Array` and applying `f` to the data.
```

Reply:
Shouldn't we mention that it *just* gives the `filenames`?

> Loads the data associated with each row of `collection` into an `Array`

PS: This sounds like we are loading the images in an array

Reply:
Ah, I was mistaken, I though based off the previous docstrings that this was `arrays`. Yes it should just say it returns the filenames, which wasn't clear before.

Reply:
> Yes it should just say it returns the filename
This will not return filenames always, I mean it can return `CCDData` if we do:
```
filenames(collection) do name
    CCDData(name)
end
```

Reply:
careful, that's not what `filenames` returns. `filenames` actually returns an iterator of strings, roughly. The strings are the filenames. In the example you showed, the return value is a consequence of using this iterator, but the iterator itself is still over the filenames. Hopefully that makes some sense.

Reply:
What would be a nice way of mentioning this in docs? i.e. it returns filenames but at the end it actually outputs an `Array` of `f` (function) operated on each filename.

Reply:
Well, that's still not true. Imagine the user creates a function that returns nothing. The output will then be a bunch of nothings! The output is in the users hands, so we shouldn't mention what the output is, or else we stand a chance of being wrong. All we can (and should) say is that it will apply the user function to the filename at each iteration

Reply:
Perfect, I will modify the docs!

Reply:
The point I'm trying to make is that all we need to describe is what makes this different from `arrays` and `ccds`. So instead of saying "this returns a vector of arrays" which is talking about the output of the user's own function, we should emphasize that this iterates and provides filenames. Anything beyond that is related to the user's function, and I don't think we should try writing documentation based on something we can't predict

Reply:
Pushed the changes!